# Necessary Imports, Add imports here as required

In [89]:
import numpy as np
import pandas as pd
import requests
import json
import collections
import random
from datetime import date
from heapq import nlargest
from zipfile import ZipFile
from dotenv import load_dotenv
from panel.interact import interact
from functools import reduce
import panel as pn
from panel import widgets
import plotly.express as px
pn.extension("plotly")
import hvplot.pandas
import kaggle as kg
import os
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
load_dotenv()
map_box_api = os.getenv("MAPBOX")

# Set the Mapbox API
px.set_mapbox_access_token(map_box_api)

# Kaggle Dataset Imports via our kaggle api

1. CO2 GHG Emissions Data
2. Greenhouse Gas Emissisons
3. Global CO2 Emissions
4. Climate Change: Earth Surface Temperature Data
5. Global Commodity Trade Statistics
6. World Development Indicators by Country

In [9]:
# !kaggle datasets download -d yoannboyere/co2-ghg-emissionsdata


  0%|          | 0.00/147k [00:00<?, ?B/s]
100%|##########| 147k/147k [00:00<00:00, 21.5MB/s]


### Process to export zip file and content

# 1. CO2 GHG Emissisons

In [10]:
# file_name = 'co2-ghg-emissionsdata.zip' #the file is your dataset exact name
# with ZipFile(file_name, 'r') as zip:
#   zip.extractall()
#   print('Done')

Done


# 2. Greenhouse Gas Emissions

In [14]:
# !kaggle datasets download -d sansuthi/greenhouse-gas-emissions


  0%|          | 0.00/567 [00:00<?, ?B/s]
100%|##########| 567/567 [00:00<00:00, 283kB/s]


In [15]:
# file_name = 'greenhouse-gas-emissions.zip' #the file is your dataset exact name
# with ZipFile(file_name, 'r') as zip:
#   zip.extractall()
#   print('Done')

Done


# 3. Global CO2 Emissions

In [16]:
# !kaggle datasets download -d sansuthi/global-co2-emissions


  0%|          | 0.00/5.90k [00:00<?, ?B/s]
100%|##########| 5.90k/5.90k [00:00<00:00, 753kB/s]


In [17]:
# file_name = 'global-co2-emissions.zip' #the file is your dataset exact name
# with ZipFile(file_name, 'r') as zip:
#   zip.extractall()
#   print('Done')

Done


# 4. Climate Change: Earth Surface Temperature Data

In [18]:
# !kaggle datasets download -d berkeleyearth/climate-change-earth-surface-temperature-data


  0%|          | 0.00/84.7M [00:00<?, ?B/s]
  2%|2         | 2.00M/84.7M [00:00<00:05, 16.4MB/s]
  5%|4         | 4.00M/84.7M [00:00<00:04, 17.2MB/s]
  7%|7         | 6.00M/84.7M [00:00<00:05, 13.9MB/s]
  9%|9         | 8.00M/84.7M [00:00<00:05, 15.1MB/s]
 12%|#1        | 10.0M/84.7M [00:00<00:05, 15.5MB/s]
 14%|#4        | 12.0M/84.7M [00:00<00:04, 16.6MB/s]
 17%|#6        | 14.0M/84.7M [00:00<00:04, 15.5MB/s]
 19%|#8        | 16.0M/84.7M [00:01<00:04, 16.7MB/s]
 21%|##1       | 18.0M/84.7M [00:01<00:04, 17.3MB/s]
 25%|##4       | 21.0M/84.7M [00:01<00:03, 19.4MB/s]
 28%|##8       | 24.0M/84.7M [00:01<00:03, 20.5MB/s]
 32%|###1      | 27.0M/84.7M [00:01<00:02, 22.3MB/s]
 35%|###5      | 30.0M/84.7M [00:01<00:02, 24.5MB/s]
 39%|###8      | 33.0M/84.7M [00:01<00:02, 26.3MB/s]
 42%|####2     | 36.0M/84.7M [00:01<00:01, 27.6MB/s]
 47%|####7     | 40.0M/84.7M [00:01<00:01, 30.1MB/s]
 52%|#####1    | 44.0M/84.7M [00:02<00:01, 29.4MB/s]
 57%|#####6    | 48.0M/84.7M [00:02<00:01, 32.4MB/s]
 

In [19]:
# file_name = 'climate-change-earth-surface-temperature-data.zip' #the file is your dataset exact name
# with ZipFile(file_name, 'r') as zip:
#   zip.extractall()
#   print('Done')

Done


# 5. Global Commodity Trade Statistics

In [20]:
# !kaggle datasets download -d unitednations/global-commodity-trade-statistics


  0%|          | 0.00/121M [00:00<?, ?B/s]
  1%|          | 1.00M/121M [00:00<00:13, 9.53MB/s]
  2%|2         | 3.00M/121M [00:00<00:08, 15.2MB/s]
  4%|4         | 5.00M/121M [00:00<00:08, 14.9MB/s]
  6%|5         | 7.00M/121M [00:00<00:07, 16.5MB/s]
  8%|8         | 10.0M/121M [00:00<00:07, 16.5MB/s]
 11%|#         | 13.0M/121M [00:00<00:06, 18.6MB/s]
 13%|#3        | 16.0M/121M [00:00<00:05, 18.4MB/s]
 16%|#5        | 19.0M/121M [00:01<00:05, 20.1MB/s]
 18%|#8        | 22.0M/121M [00:01<00:04, 22.0MB/s]
 21%|##        | 25.0M/121M [00:01<00:04, 23.3MB/s]
 23%|##3       | 28.0M/121M [00:01<00:04, 23.8MB/s]
 27%|##6       | 32.0M/121M [00:01<00:03, 26.0MB/s]
 29%|##9       | 35.0M/121M [00:01<00:03, 27.3MB/s]
 32%|###2      | 39.0M/121M [00:01<00:02, 29.5MB/s]
 36%|###5      | 43.0M/121M [00:01<00:02, 31.2MB/s]
 39%|###8      | 47.0M/121M [00:02<00:02, 30.8MB/s]
 41%|####1     | 50.0M/121M [00:02<00:02, 30.5MB/s]
 45%|####4     | 54.0M/121M [00:02<00:02, 32.7MB/s]
 48%|####8     | 58.

In [21]:
# file_name = 'global-commodity-trade-statistics.zip' #the file is your dataset exact name
# with ZipFile(file_name, 'r') as zip:
#   zip.extractall()
#   print('Done')

Done


# 6. World Development Indicators By Country

In [22]:
# !kaggle datasets download -d hn4ever/world-development-indicators-by-countries


  0%|          | 0.00/51.9k [00:00<?, ?B/s]
100%|##########| 51.9k/51.9k [00:00<00:00, 17.7MB/s]


In [23]:
# file_name = 'world-development-indicators-by-countries.zip' #the file is your dataset exact name
# with ZipFile(file_name, 'r') as zip:
#   zip.extractall()
#   print('Done')

Done


# Bringing the Data In 

In [57]:
carbon_emission_data = pd.read_csv('datasets/co2_emission.csv')
carbon_emission_life_expect_data =  pd.read_csv('datasets/CO2Emission_LifeExp.csv')
trade_statistics =  pd.read_csv('datasets/commodity_trade_statistics_data.csv', low_memory=False)
ghg_by_sector_data =  pd.read_csv('datasets/GHG-Emissions-by-sector.csv')
land_temp_by_city_data =  pd.read_csv('datasets/GlobalLandTemperaturesByCity.csv')
ghg_emissions_data = pd.read_csv('datasets/greenhouse_gas_emissions.csv')
sustainability_data = pd.read_csv('datasets/sustainability.csv')
energy_data = pd.read_csv('datasets/energy.csv')

# Data Exploration

# Carbon Emission Data Cleaning

In [58]:
carbon_emission_data

Entity Code  Year  Annual CO₂ emissions (tonnes )
0      Afghanistan  AFG  1949                        14656.00
1      Afghanistan  AFG  1950                        84272.00
2      Afghanistan  AFG  1951                        91600.00
3      Afghanistan  AFG  1952                        91600.00
4      Afghanistan  AFG  1953                       106256.00
...            ...  ...   ...                             ...
20848     Zimbabwe  ZWE  2013                     11536239.29
20849     Zimbabwe  ZWE  2014                     11866348.41
20850     Zimbabwe  ZWE  2015                     10907603.94
20851     Zimbabwe  ZWE  2016                      9932649.88
20852     Zimbabwe  ZWE  2017                     10397718.47

[20853 rows x 4 columns]

In [110]:
pivoted_data = carbon_emission_data.pivot(index='Entity', columns=['Year'], values='Annual CO₂ emissions (tonnes )')
pivoted_data.head()

Year              1751  1752  1753  1754  1755  1756  1757  1758  1759  1760  \
Entity                                                                         
Afghanistan        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Africa             0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
Albania            NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Algeria            NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Americas (other)   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   

Year              ...          2008          2009          2010          2011  \
Entity            ...                                                           
Afghanistan       ...  4.198330e+06  6.760313e+06  8.452224e+06  1.222602e+07   
Africa            ...  1.178143e+09  1.188276e+09  1.193784e+09  1.198373e+09   
Albania           ...  4.371152e+06  4.374816e+06  4.594656e+06  5.235856e+06   
Algeria           ...  1.090318e+08  1.201859e+08  1.178141e+08  1.198069e+08   
Americas (other)  ...  2.226580e+09  2.157619e+09  2.253919e+09  2.294779e+09   

Year                      2012          2013          2014          2015  \
Entity                                                                     
Afghanistan       1.074206e+07  1.000041e+07  9.789186e+06  1.010743e+07   
Africa            1.233842e+09  1.250644e+09  1.290489e+09  1.289003e+09   
Albania           4.906096e+06  5.059984e+06  5.712176e+06  5.947724e+06   
Algeria           1.281100e+08  1.324346e+08  1.431771e+08  1.505925e+08   
Americas (other)  2.375406e+09  2.406775e+09  2.443171e+09  2.435554e+09   

Year                      2016          2017  
Entity                                        
Afghanistan       1.225791e+07  1.301474e+07  
Africa            1.310656e+09  1.331740e+09  
Albania           6.272858e+06  6.379207e+06  
Algeria           1.503780e+08  1.505514e+08  
Americas (other)  2.415058e+09  2.398632e+09  

[5 rows x 267 columns]

In [118]:
cleaned_country_data = pivoted_data.rename_axis(None)

In [124]:
cleaned_country_data_transposed = cleaned_country_data.T

In [125]:
cleaned_country_data_transposed

Afghanistan        Africa     Albania      Algeria  Americas (other)  \
Year                                                                         
1751          NaN  0.000000e+00         NaN          NaN      0.000000e+00   
1752          NaN  0.000000e+00         NaN          NaN      0.000000e+00   
1753          NaN  0.000000e+00         NaN          NaN      0.000000e+00   
1754          NaN  0.000000e+00         NaN          NaN      0.000000e+00   
1755          NaN  0.000000e+00         NaN          NaN      0.000000e+00   
...           ...           ...         ...          ...               ...   
2013  10000414.29  1.250644e+09  5059984.00  132434630.3      2.406775e+09   
2014   9789185.57  1.290489e+09  5712176.00  143177050.1      2.443171e+09   
2015  10107428.68  1.289003e+09  5947724.28  150592497.7      2.435554e+09   
2016  12257907.69  1.310656e+09  6272857.94  150378025.5      2.415058e+09   
2017  13014739.66  1.331740e+09  6379207.00  150551376.6      2.398632e+09   

        Andorra       Angola   Anguilla  Antarctic Fisheries  \
Year                                                           
1751        NaN          NaN        NaN                  NaN   
1752        NaN          NaN        NaN                  NaN   
1753        NaN          NaN        NaN                  NaN   
1754        NaN          NaN        NaN                  NaN   
1755        NaN          NaN        NaN                  NaN   
...         ...          ...        ...                  ...   
2013  476320.00  32373114.67  135568.00                  NaN   
2014  461664.00  34457641.91  142896.00                  NaN   
2015  477238.62  34761128.89  149176.85                  NaN   
2016  511506.85  35351002.67  152072.24                  NaN   
2017  520183.81  35389928.27  152045.24                  NaN   

      Antigua and Barbuda  ...     Uruguay    Uzbekistan    Vanuatu  \
Year                       ...                                        
1751                  NaN  ...         NaN           NaN        NaN   
1752                  NaN  ...         NaN           NaN        NaN   
1753                  NaN  ...         NaN           NaN        NaN   
1754                  NaN  ...         NaN           NaN        NaN   
1755                  NaN  ...         NaN           NaN        NaN   
...                   ...  ...         ...           ...        ...   
2013            523952.00  ...  7489843.70  1.026549e+08  106256.00   
2014            531280.00  ...  6656106.94  1.044808e+08  153888.00   
2015            554631.88  ...  6651105.74  1.106693e+08  167934.09   
2016            565396.80  ...  6764424.63  9.792527e+07  169858.44   
2017            565296.40  ...  6840339.75  9.899895e+07  181363.53   

        Venezuela      Vietnam  Wallis and Futuna Islands         World  \
Year                                                                      
1751          NaN          NaN                        NaN  9.350528e+06   
1752          NaN          NaN                        NaN  9.354192e+06   
1753          NaN          NaN                        NaN  9.354192e+06   
1754          NaN          NaN                        NaN  9.357856e+06   
1755          NaN          NaN                        NaN  9.361520e+06   
...           ...          ...                        ...           ...   
2013  182868919.2  140438875.8                   21984.00  3.520789e+10   
2014  184236581.7  159712377.1                   21984.00  3.550583e+10   
2015  176897278.5  184501879.2                   23990.58  3.546275e+10   
2016  166096367.9  197801557.3                   24265.49  3.567510e+10   
2017  159567635.4  198826548.8                   25909.08  3.615326e+10   

            Yemen      Zambia     Zimbabwe  
Year                                        
1751          NaN         NaN          NaN  
1752          NaN         NaN          NaN  
1753          NaN         NaN          NaN  
1754          NaN         NaN    

In [126]:
global_carbon_emissions = cleaned_country_data_transposed['World']

In [133]:
global_carbon_emissions.hvplot.line(min_width=1000,height=500)

:Curve   [Year]   (World)

In [128]:
global_carbon_emissions_ex_total = cleaned_country_data_transposed.drop('World', axis=1)

In [137]:
global_carbon_emissions_ex_total.hvplot.line(width=1500,height=500)

:NdOverlay   [Variable]
   :Curve   [Year]   (value)

# Grouping By Country for individual country analysis and statistics

In [59]:
country_groups = carbon_emission_data.groupby('Entity')

In [60]:
country_dataframe_list = [country_groups.get_group(x) for x in country_groups.groups]

In [63]:
country_dataframe_list[0].rename(columns={'Annual CO₂ emissions (tonnes )':country_dataframe_list[0]['Entity'][0]})

Entity Code  Year  Afghanistan
0   Afghanistan  AFG  1949     14656.00
1   Afghanistan  AFG  1950     84272.00
2   Afghanistan  AFG  1951     91600.00
3   Afghanistan  AFG  1952     91600.00
4   Afghanistan  AFG  1953    106256.00
..          ...  ...   ...          ...
64  Afghanistan  AFG  2013  10000414.29
65  Afghanistan  AFG  2014   9789185.57
66  Afghanistan  AFG  2015  10107428.68
67  Afghanistan  AFG  2016  12257907.69
68  Afghanistan  AFG  2017  13014739.66

[69 rows x 4 columns]

In [116]:
for country in country_dataframe_list:
    country.rename(columns={'Annual CO₂ emissions (tonnes )':country.iloc[0]['Entity']}, inplace=True)

In [117]:
country_dataframe_list[0]

Entity Code  Year  Afghanistan
0   Afghanistan  AFG  1949     14656.00
1   Afghanistan  AFG  1950     84272.00
2   Afghanistan  AFG  1951     91600.00
3   Afghanistan  AFG  1952     91600.00
4   Afghanistan  AFG  1953    106256.00
..          ...  ...   ...          ...
64  Afghanistan  AFG  2013  10000414.29
65  Afghanistan  AFG  2014   9789185.57
66  Afghanistan  AFG  2015  10107428.68
67  Afghanistan  AFG  2016  12257907.69
68  Afghanistan  AFG  2017  13014739.66

[69 rows x 4 columns]

In [20]:
unique_countries = list(carbon_emission_data['Entity'].unique())

In [21]:
for country in unique_countries:
    print(country)

Afghanistan
Africa
Albania
Algeria
Americas (other)
Andorra
Angola
Anguilla
Antarctic Fisheries
Antigua and Barbuda
Argentina
Armenia
Aruba
Asia and Pacific (other)
Australia
Austria
Azerbaijan
Bahamas
Bahrain
Bangladesh
Barbados
Belarus
Belgium
Belize
Benin
Bermuda
Bhutan
Bolivia
Bonaire Sint Eustatius and Saba
Bosnia and Herzegovina
Botswana
Brazil
British Virgin Islands
Brunei
Bulgaria
Burkina Faso
Burundi
Cambodia
Cameroon
Canada
Cape Verde
Cayman Islands
Central African Republic
Chad
Chile
China
Christmas Island
Colombia
Comoros
Republic of the Congo
Cook Islands
Costa Rica
Cote d'Ivoire
Croatia
Cuba
Curacao
Cyprus
Czech Republic
Czechoslovakia
Democratic Republic of Republic of the Congo
Denmark
Djibouti
Dominica
Dominican Republic
EU-28
Ecuador
Egypt
El Salvador
Equatorial Guinea
Eritrea
Estonia
Ethiopia
Europe (other)
Faeroe Islands
Falkland Islands
Fiji
Finland
France
French Guiana
French Polynesia
Gabon
Gambia
Georgia
Germany
Ghana
Gibraltar
Greece
Greenland
Grenada
Guadeloup

In [31]:
carbon_emission_data['Entity'].nunique()

233

# Our Dataset has 233 Unique Countries with ESG Data to explore. We Shall be looking at the correlation between country sub industry and Carbon and GHG emissions.

In [6]:
carbon_emission_life_expect_data

Country Code  CO2Emissions  YearlyChange  Percapita  Population  \
0       Afghanistan  AFG       9900004          7.13       0.28    35383032   
1           Albania  ALB       5208319          4.45       1.80     2886438   
2           Algeria  DZA     156220560          0.17       3.85    40551392   
3            Angola  AGO      30566933          3.13       1.06    28842489   
4          Anguilla  AIA         30262          1.52       2.10       14429   
..              ...  ...           ...           ...        ...         ...   
203         Vietnam  VNM     206042140          0.09       2.20    93640422   
204  Western Sahara  ESH        207592          2.61       0.39      538749   
205           Yemen  YEM      25647990          1.62       0.94    27168208   
206          Zambia  ZMB       4230584          3.67       0.26    16363458   
207        Zimbabwe  ZWE      10062628         -4.17       0.72    14030331   

     LifeExpectancy  
0            63.763  
1            78.194  
2            76.298  
3            59.925  
4            81.441  
..              ...  
203          75.172  
204          69.496  
205          66.087  
206          62.464  
207          60.294  

[208 rows x 7 columns]

In [7]:
trade_statistics

country_or_area  year comm_code  \
0           Afghanistan  2016     10410   
1           Afghanistan  2016     10420   
2           Afghanistan  2008     10210   
3               Albania  2016     10290   
4               Albania  2016     10392   
...                 ...   ...       ...   
8225866        Zimbabwe  2001     TOTAL   
8225867        Zimbabwe  2001     TOTAL   
8225868        Zimbabwe  2000     TOTAL   
8225869        Zimbabwe  1995     TOTAL   
8225870        Zimbabwe  1995     TOTAL   

                                               commodity       flow  \
0                                            Sheep, live     Export   
1                                            Goats, live     Export   
2                Bovine animals, live pure-bred breeding     Import   
3        Bovine animals, live, except pure-bred breeding     Import   
4          Swine, live except pure-bred breeding > 50 kg     Import   
...                                                  ...        ...   
8225866                                  ALL COMMODITIES     Export   
8225867                                  ALL COMMODITIES  Re-Export   
8225868                                  ALL COMMODITIES     Export   
8225869                                  ALL COMMODITIES     Import   
8225870                                  ALL COMMODITIES     Export   

          trade_usd  weight_kg    quantity_name  quantity         category  
0              6088     2339.0  Number of items      51.0  01_live_animals  
1              3958      984.0  Number of items      53.0  01_live_animals  
2           1026804      272.0  Number of items    3769.0  01_live_animals  
3           2414533  1114023.0  Number of items    6853.0  01_live_animals  
4          14265937  9484953.0  Number of items   96040.0  01_live_animals  
...             ...        ...              ...       ...              ...  
8225866  1206807424        NaN      No Quantity       NaN  all_commodities  
8225867    15943977        NaN      No Quantity       NaN  all_commodities  
8225868  1924962432        NaN      No Quantity       NaN  all_commodities  
8225869  2658853376        NaN      No Quantity       NaN  all_commodities  
8225870  1845905792        NaN      No Quantity       NaN  all_commodities  

[8225871 rows x 10 columns]

In [10]:
ghg_by_sector_data

Sub-sector  GlobalShare
0                                    Road         11.9
1                                Aviation          1.9
2                                    Rail          0.4
3                                Pipeline          0.3
4                                    Ship          1.7
5                             Residential         10.9
6                              Commercial          6.6
7                            Iron & Steel          7.2
8                       Non-ferous metals          0.7
9                               Machinery          0.5
10                       Food and tobacco          1.0
11                  Paper pulp & printing          0.6
12      Chemical & petrochemical (energy)          3.6
13                         Other industry         10.6
14               Energy in Agri & Fishing          1.7
15            Unallocated fuel combustion          7.8
16                                   Coal          1.9
17                      Oil & Natural Gas          3.9
18                                 Cement          3.0
19  Chemical & petrochemical (industrial)          2.2
20                     Livestock & Manure          5.8
21                       Rice Cultivation          1.3
22                     Agricultural Soils          4.1
23                           Crop Burning          3.5
24                            Forest Land          2.2
25                               Cropland          1.4
26                              Grassland          0.1
27                              Landfills          1.9
28                             Wastewater          1.3

In [12]:
land_temp_by_city_data

dt  AverageTemperature  AverageTemperatureUncertainty  \
0        1743-11-01               6.068                          1.737   
1        1743-12-01                 NaN                            NaN   
2        1744-01-01                 NaN                            NaN   
3        1744-02-01                 NaN                            NaN   
4        1744-03-01                 NaN                            NaN   
...             ...                 ...                            ...   
8599207  2013-05-01              11.464                          0.236   
8599208  2013-06-01              15.043                          0.261   
8599209  2013-07-01              18.775                          0.193   
8599210  2013-08-01              18.025                          0.298   
8599211  2013-09-01                 NaN                            NaN   

           City      Country Latitude Longitude  
0         Århus      Denmark   57.05N    10.33E  
1         Århus      Denmark   57.05N    10.33E  
2         Århus      Denmark   57.05N    10.33E  
3         Århus      Denmark   57.05N    10.33E  
4         Århus      Denmark   57.05N    10.33E  
...         ...          ...      ...       ...  
8599207  Zwolle  Netherlands   52.24N     5.26E  
8599208  Zwolle  Netherlands   52.24N     5.26E  
8599209  Zwolle  Netherlands   52.24N     5.26E  
8599210  Zwolle  Netherlands   52.24N     5.26E  
8599211  Zwolle  Netherlands   52.24N     5.26E  

[8599212 rows x 7 columns]

#Instructor Notes
1. What about other potentioal industries that they can expand on if other industries are set to perform poorly or die if there is a reduction impact
2. What can be some of the positives that can be create dout of a reduction
3. What doe si tmean for a potential growth point of view for ht ecountries that stand to lose from CO2 reduction policies
4. Counteract the negatives of the loss with the positives of a gain in green house gas reductions and co2 reductions
5. We will narrow to conutries that have the most impact 
6. Look at th elife expectancy and the correlationw ith a reduction or increase in carbon emissions
7. Look at regional impacts 